File: src/api_server/ccapi/actions/v3/host/usercustom.go
/*
 * Tencent is pleased to support the open source community by making 蓝鲸 available.
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except 
 * in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and 
 * limitations under the License.
 */
 
package host

import (
	"configcenter/src/api_server/ccapi/actions/v3"
	"configcenter/src/common"
	"configcenter/src/common/core/cc/actions"
	"configcenter/src/common/core/cc/api"
	httpcli "configcenter/src/common/http/httpclient"
	"io"

	restful "github.com/emicklei/go-restful"
)

// SaveUserCustom save user configure
func SaveUserCustom(req *restful.Request, resp *restful.Response) {
	cc := api.NewAPIResource()
	url := cc.HostAPI() + "/host/v1/usercustom"
	rsp, _ := httpcli.ReqForward(req, url, common.HTTPCreate)
	io.WriteString(resp, rsp)
}

// GetUser get useer configure
func GetUser(req *restful.Request, resp *restful.Response) {
	cc := api.NewAPIResource()
	url := cc.HostAPI() + "/host/v1/usercustom/user/search"
	rsp, _ := httpcli.ReqForward(req, url, common.HTTPSelectPost)
	io.WriteString(resp, rsp)
}

// GetDefault get user default configure
func GetDefault(req *restful.Request, resp *restful.Response) {
	cc := api.NewAPIResource()
	url := cc.HostAPI() + "/host/v1/usercustom/default/search"
	rsp, _ := httpcli.ReqForward(req, url, common.HTTPSelectPost)
	io.WriteString(resp, rsp)
}

func init() {
	actions.RegisterNewAction(actions.Action{Verb: common.HTTPCreate, Path: "/usercustom", Params: nil, Handler: SaveUserCustom, Version: v3.APIVersion})
	actions.RegisterNewAction(actions.Action{Verb: common.HTTPSelectPost, Path: "/usercustom/user/search", Params: nil, Handler: GetUser, Version: v3.APIVersion})
	actions.RegisterNewAction(actions.Action{Verb: common.HTTPSelectPost, Path: "/usercustom/default/search", Params: nil, Handler: GetDefault, Version: v3.APIVersion})

}

File: src/ui/src/store/modules/api/user-custom.js
/*
 * Tencent is pleased to support the open source community by making 蓝鲸 available.
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

import $http from '@/api'

const state = {
    usercustom: {}
}

const getters = {
    classifyNavigationKey: (state, getters, rootState, rootGetters) => {
        const bizId = rootGetters['objectBiz/bizId']
        const isAdminView = rootGetters['isAdminView']
        const userName = rootGetters['userName']
        return `${userName}_${isAdminView ? 'adminView' : bizId}_classify_navigation`
    },
    firstEntryKey: (state, getters, rootState, rootGetters) => {
        const bizId = rootGetters['objectBiz/bizId']
        const isAdminView = rootGetters['isAdminView']
        const userName = rootGetters['userName']
        return `${userName}_${isAdminView ? 'adminView' : bizId}_first_entry`
    },
    recentlyKey: (state, getters, rootState, rootGetters) => {
        const bizId = rootGetters['objectBiz/bizId']
        const isAdminView = rootGetters['isAdminView']
        const userName = rootGetters['userName']
        return `${userName}_${isAdminView ? 'adminView' : bizId}_recently`
    },
    usercustom: state => state.usercustom,
    getCustomData: (state) => (key, defaultData = null) => {
        if (state.usercustom.hasOwnProperty(key)) {
            return state.usercustom[key]
        }
        return defaultData
    }
}

const actions = {
    /**
     * 保存用户字段配置
     * @param {Function} commit store commit mutation hander
     * @param {Object} state store state
     * @param {String} dispatch store dispatch action hander
     * @param {Object} params 参数
     * @return {promises} promises 对象
     */
    saveUsercustom ({ commit, state, dispatch }, usercustom = {}) {
        return $http.post(`usercustom`, usercustom, { cancelWhenRouteChange: false }).then(() => {
            $http.cancelCache('searchUserCustom')
            commit('setUsercustom', usercustom)
            return state.usercustom
        })
    },

    /**
     * 获取用户字段配置
     * @param {Function} commit store commit mutation hander
     * @param {Object} state store state
     * @param {String} dispatch store dispatch action hander
     * @return {promises} promises 对象
     */
    searchUsercustom ({ commit, state, dispatch }, { config }) {
        const mergedConfig = Object.assign({
            requestId: 'searchUserCustom'
        }, config)
        return $http.post(`usercustom/user/search`, {}, mergedConfig).then(usercustom => {
            commit('setUsercustom', usercustom)
            return usercustom
        })
    },

    /**
     * 获取默认字段配置
     * @param {Function} commit store commit mutation hander
     * @param {Object} state store state
     * @param {String} dispatch store dispatch action hander
     * @return {promises} promises 对象
     */
    getUserDefaultCustom ({ commit, state, dispatch }) {
        return $http.post(`usercustom/default/search`)
    },

    setRencentlyData ({ commit, state, dispatch }, { id }) {
        const usercustomData = state.usercustom.recently_models || []
        const isExist = usercustomData.some(target => target === id)
        let newUsercustomData = [...usercustomData]
        if (isExist) {
            newUsercustomData = newUsercustomData.filter(target => target !== id)
        }
        newUsercustomData.unshift(id)
        dispatch('saveUsercustom', {
            recently_models: newUsercustomData
        })
    }
}

const mutations = {
    setUsercustom (state, usercustom) {
        state.usercustom = Object.assign({}, state.usercustom, usercustom)
    }
}

export default {
    namespaced: true,
    state,
    getters,
    actions,
    mutations
}
